crates/red_knot_module_resolver/src/path.rs

```diff
@@ -725,6 +707,169 @@ impl PartialEq<ModuleResolutionPathRef<'_>> for VendoredPathBuf {
     }
 }
 
+#[derive(Debug, PartialEq, Eq)]
+pub enum SearchPathValidationError {
+    NotADirectory(SystemPathBuf),
+    NoStdlibSubdirectory(SystemPathBuf),
+    NoVersionsFile(SystemPathBuf),
+    VersionsParseError(TypeshedVersionsParseError),
+}
+
+impl fmt::Display for SearchPathValidationError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NotADirectory(path) => write!(f, "{path} does not point to a directory"),
+            Self::NoStdlibSubdirectory(path) => {
+                write!(f, "The directory at {path} has no `stdlib/` subdirectory")
+            }
+            Self::NoVersionsFile(path) => write!(f, "Expected a file at {path}/stldib/VERSIONS"),
+            Self::VersionsParseError(underlying_error) => underlying_error.fmt(f),
+        }
+    }
+}
+
+impl std::error::Error for SearchPathValidationError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        if let Self::VersionsParseError(underlying_error) = self {
+            Some(underlying_error)
+        } else {
+            None
+        }
+    }
+}
+
+type SearchPathResult = Result<ModuleSearchPath, SearchPathValidationError>;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub(crate) struct ModuleSearchPath(Arc<ModuleResolutionPathBuf>);
+
+impl ModuleSearchPath {
+    pub(crate) fn extra(system: &dyn System, root: impl Into<SystemPathBuf>) -> SearchPathResult {
+        let root = root.into();
+        if system.is_directory(&root) {
+            Ok(Self(Arc::new(ModuleResolutionPathBuf::extra(
+                SystemPath::absolute(root, system.current_directory()),
+            ))))
+        } else {
+            Err(SearchPathValidationError::NotADirectory(root))
+        }
+    }
+
+    pub(crate) fn first_party(
+        system: &dyn System,
+        root: impl Into<SystemPathBuf>,
+    ) -> SearchPathResult {
+        let root = root.into();
+        if system.is_directory(&root) {
+            Ok(Self(Arc::new(ModuleResolutionPathBuf::first_party(
+                SystemPath::absolute(root, system.current_directory()),
+            ))))
+        } else {
+            Err(SearchPathValidationError::NotADirectory(root))
+        }
+    }
+
+    pub(crate) fn custom_stdlib(
+        db: &dyn Db,
+        typeshed: impl Into<SystemPathBuf>,
+    ) -> SearchPathResult {
+        let typeshed = typeshed.into();
+        let system = db.system();
+        if !system.is_directory(&typeshed) {
+            return Err(SearchPathValidationError::NotADirectory(typeshed));
+        }
+        let stdlib = typeshed.join("stdlib");
+        if !system.is_directory(&stdlib) {
```

Comment:
I thought search-path settings were set once at the start of the process, and then never changed after that. Given that, is Salsa invalidation necessary here? If the settings never change, then the settings will only ever be validated once, so this function will only ever be executed once.

Reply:
I think we need to watch these paths and if something we are assuming is a directory becomes not-a-directory, we need to react to that -- but I'd also assumed that this type of change (which should be rare) would mean totally starting over, which seems like it means Salsa invalidation wouldn't be relevant.

Reply:
Watching the directories wouldn't be sufficient. We would have to watch the parent directories because a change from file to directory is technically a different file (watching is file and not path-based). But that would significantly increase the number of files we watch, and some systems limit the number of files that can be watched. That's why I think that we don't want to do that.

There's also the problem that we can't watch the folders if resolving the module resolution failed. At least not without duplicating a lot of this logic in the code where  we set-up file watching. That's why I'm actually not sure if we want to error in the first place or if it would be better fallback to reasonable defaults. Because failing to start means that the LSP can support no-completion whatsoever. It also can't provide any help fixing the ruff settings. It's just dead. 

> I thought search-path settings were set once at the start of the process, and then never changed after that.

This is correct today but not long-term. A user can change the workspace settings at any time, and we have to respond to that.

Reply:
> > I thought search-path settings were set once at the start of the process, and then never changed after that.
> 
> This is correct today but not long-term. A user can change the workspace settings at any time, and we have to respond to that.

I see -- this is different to what I had previously understood